### PR TITLE
check for unused bindings

### PR DESCRIPTION
**Pull request by ZacLN**
Needs some finessing:
- [x] special handling for loops 
- [x] do backwards/forward looks for overwritten bindings

### CODE REVIEW

File: src/linting/checks.jl [hunk 1]
```diff
@@ -860,3 +860,132 @@ function check_const(x::EXPR)
...
+all_underscore(s::String) = all(==(0x5f), codeunits(s))
+
+function is_sig_arg(x)
+    is_in_fexpr(x, CSTParser.iscall)
+end
+
+function is_overwritten_in_loop(x)
+    # Cuts out false positives for check_unused_binding - the linear nature of our 
+    # semantic passes mean a variable declared at the end of a loop's block but used at 
+    # the start won't appear to be referenced.
```
Comment (pfitzseb) [inline src/linting/checks.jl:882]: Hm, but isn't that in line with how Julia's loops work?
```
julia> for i in 1:3
         if i > 2
           println(x)
         end
         x = i
       end
ERROR: UndefVarError: x not defined
```
So we only need to check if the variable being assigned is used in a enclosing scope, as you mention below?

Comment (ZacLN) [reply src/linting/checks.jl:882]: Hmm, that i did not know!

Comment (pfitzseb) [reply src/linting/checks.jl:885]: Doesn't matter whether it's in a top-level scope or not, afaict:
```
julia> function f()
         for i in 1:3
           if i > 2
             println(x)
           end
           x=i
         end
       end
f (generic function with 1 method)

julia> f()
ERROR: UndefVarError: x not defined
```

Comment (ZacLN) [reply src/linting/checks.jl:885]: I meant for the StaticLint checks

**Review by pfitzseb — APPROVED** [review]

**Pull request merged**